src/waveform/renderers/allshader/waveformrenderertextured.cpp

```diff
@@ -0,0 +1,515 @@
+        m_frameShaderProgram = std::make_unique<QOpenGLShaderProgram>();
+    }
+
+    if (!loadShaders()) {
+        return;
+    }
+    createFrameBuffers();
+    createGeometry();
+    if (!loadTexture()) {
+        return;
+    }
+}
+
+void WaveformRendererTextured::onSetup(const QDomNode& node) {
+    Q_UNUSED(node);
+}
+
+void WaveformRendererTextured::onSetTrack() {
+    if (m_loadedTrack) {
+        disconnect(m_loadedTrack.get(),
+                &Track::waveformUpdated,
+                this,
+                &WaveformRendererTextured::slotWaveformUpdated);
+    }
+
+    slotWaveformUpdated();
+
+    const TrackPointer pTrack = m_waveformRenderer->getTrackInfo();
+    if (!pTrack) {
+        return;
+    }
+
+    // When the track's waveform has been changed (or cleared), it is necessary
+    // to update (or delete) the texture containing the waveform which was
+    // uploaded to GPU. Otherwise, previous waveform will be shown.
+    connect(pTrack.get(),
+            &Track::waveformUpdated,
+            this,
+            &WaveformRendererTextured::slotWaveformUpdated);
+
+    m_loadedTrack = pTrack;
+}
+
+void WaveformRendererTextured::resizeGL(int, int) {
+    createFrameBuffers();
+}
+
+void WaveformRendererTextured::slotWaveformUpdated() {
+    m_textureRenderedWaveformCompletion = 0;
+    // initializeGL not called yet
+    if (!m_frameShaderProgram) {
+        return;
+    }
+    loadTexture();
+}
+
+void WaveformRendererTextured::paintGL() {
+    ConstWaveformPointer pWaveform = m_waveformRenderer->getWaveform();
+    if (pWaveform.isNull()) {
+        return;
+    }
+
+    const double audioVisualRatio = pWaveform->getAudioVisualRatio();
+    if (audioVisualRatio <= 0) {
+        return;
+    }
+
+    int dataSize = pWaveform->getDataSize();
+    if (dataSize <= 1) {
+        return;
+    }
+
+    const WaveformData* data = pWaveform->data();
+    if (data == nullptr) {
+        return;
+    }
+
+    const double trackSamples = m_waveformRenderer->getTrackSamples();
+    if (trackSamples <= 0) {
+        return;
+    }
+
+    // NOTE(vRince): completion can change during loadTexture
+    // do not remove currenCompletion temp variable !
+    const int currentCompletion = pWaveform->getCompletion();
+    if (m_textureRenderedWaveformCompletion < currentCompletion) {
+        loadTexture();
+        m_textureRenderedWaveformCompletion = currentCompletion;
+    }
+
+    // Per-band gain from the EQ knobs.
+    float lowGain(1.0), midGain(1.0), highGain(1.0), allGain(1.0);
+    getGains(&allGain, true, &lowGain, &midGain, &highGain);
+
+    const auto firstVisualIndex = static_cast<GLfloat>(
+            m_waveformRenderer->getFirstDisplayedPosition() * trackSamples /
+            audioVisualRatio / 2.0);
+    const auto lastVisualIndex = static_cast<GLfloat>(
+            m_waveformRenderer->getLastDisplayedPosition() * trackSamples / audioVisualRatio / 2.0);
+
+    // const int firstIndex = int(firstVisualIndex+0.5);
```

Comment:
remove comment cruft?

Reply:
I just copied this as-is from the legacy code.